L50-objc-interceptor.md

```diff
@@ -0,0 +1,263 @@
+// this array. This parameter should not be modified by any interceptor and will
+// not take effect if done so.
+@property(readwrite) NSArray *interceptorFactories;
+
+@end
+```
+
+### GRPCCall2 interface
+GRPCCall2, the interface that current gRPC ObjC users interacts with, becomes a simple wrapper that forwards messages to/from the first interceptor in the chain. The only additional check that we perform in a GRPCCall2 instance is a simple check of whether the chain of interceptors are already initialized.
+
+All concrete call logics are handled by GRPCCall2Internal, which is a special interceptor that do not chain to another interceptor. This strategy maintains compatibility to the current users of GRPCCall2.
+
+## Implementation
+
+The prototype of this design proposal is close to completion and will be published by @muxi soon.
+
+## Open issues
+
+### Intercept protobuf message
+The current interceptor design does not allow intercepting protobuf message. Since the interceptors are behind the GRPCCall2 object, any message seen by interceptors are raw data.
+
+This problem makes it difficult for an interceptor that would like, e.g., to change a protobuf field in the request. One solution is to have the interceptor deserialize the message in the interceptor, modify it, serialize it again, then pass it on. Such an approach works but is inefficient.
+
+A possible enhancement to solve this problem is to introduce the concept of serializer to GRPCCall2 interface. It can be an object conforming to a serializer interface and passed along with callOptions.
+
+```objectivec
+
+@protocol <GRPCMarshaller>
+
+- (NSData *)serializeData:(id)data;
+
+- (id)deserializeData:(NSData *)Data;
+
+@end
+
+@interface GRPCMutableCallOptions
+
+// Properties that already exist
+@property id<GRPCMarshaller> marshaller;
+
+@end
+```
+After all the alternations in the interceptors, GRPCCall2Internal object uses the serializer to transform the data before passing it to core.
+
+Though there are some potential use cases for this feature, it can be implemented independent of the proposal above.
+
+### Interceptor Dispatch Queues
+Since each interceptor is stateful, multiple requests or responses issued to an interceptor should be processed in order. That requires the dispatch queues of an interceptor to be serial queue. This requirement currently seems reasonable for us, because multiple calls can each have their own dispatch queue so that the same class of interceptors for different calls can still run concurrently. On the other hand, if serialized process is required across all interceptors of the same class, they will be able to use the same dispatch queue provided by their interceptor factory on initialization.
+
+In case we have to support concurrent queue in the future, the following alternation may be used to achieve the goal transparently to the users and the interceptor developers:
```

Comment:
alternation => alternative

Reply:
Well I didn't really mean "alternative" too :) But I rephrased it. Thanks Sanjay.